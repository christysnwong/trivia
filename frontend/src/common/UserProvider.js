import React, {useEffect, useState} from "react";
import UserContext from "./UserContext";

const demoUser = {
  id: 1,
  username: "testuser",
  first_name: "Test",
  last_name: "User",
  email: "test@test.com",
  stats: {
    level: 14,
    title: "Pro",
    "quizzes completed": 34,
    points: 3440,
    remainingPts: 60,
    levelPts: 300,
  },
};

const UserProvider = ({
  children,
  currUser = demoUser
}) => {
  
  const [currentUser, setCurrUser] = useState();

  return (
    // <UserContext.Provider value={{ currUser }}>
    <UserContext.Provider value={{ currUser, setCurrUser }}>
      {children}
    </UserContext.Provider>
  );
};

export default UserProvider;
