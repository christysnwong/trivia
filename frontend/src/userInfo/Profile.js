import { useContext } from "react";
import TriviaApi from "../api/api";
import UserContext from "../common/UserContext";
import { useHistory } from "react-router-dom";

import Noty from "noty";

const Profile = () => {
  const { currUser, setCurrUser, setToken } = useContext(UserContext);

  const history = useHistory();

  // remove user's account
  const removeUser = () => {
    try {
      TriviaApi.removeUser(currUser.username).then((res) => {
        if (res.deleted) {
          setCurrUser(null);
          setToken(null);
          TriviaApi.token = null;

          new Noty({
            type: "success",
            layout: "topRight",
            theme: "bootstrap-v4",
            text: "This account is successfully deleted.",
            timeout: "3000",
            progressBar: true,
            closeWith: ["click"],
            killer: true,
          }).show();

          history.push("/");
        }
      });
    } catch (e) {
      console.error(e);

      new Noty({
        type: "error",
        layout: "topRight",
        theme: "bootstrap-v4",
        text: e,
        timeout: "3000",
        progressBar: true,
        closeWith: ["click"],
        killer: true,
      }).show();

    }
    
  };

  return (
    <div className="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4">
      <h3 className="mb-5">
        <i className="fa-solid fa-address-card me-2"></i>
        Profile
      </h3>

      {Object.entries(currUser)
        .filter((entry) => entry[0] !== "id" && entry[0] !== "stats")
        .map((entry, idx) => (
          <div className="row mb-3" key={idx}>
            <div className="col-4 text-capitalize fw-bold">{entry[0]}</div>
            <div className="col-8">{entry[1]}</div>
          </div>
        ))}

      <button
        type="button"
        className="btn btn-success mt-4 me-2"
        onClick={() => history.push("/profile/edit")}
        method="get"
      >
        Edit your info
      </button>

      <button
        type="button"
        className="btn btn-danger mt-4"
        data-bs-toggle="modal"
        data-bs-target="#deleteModal"
      >
        Delete this account
      </button>

      {/* <!-- Modal --> */}
      <div
        className="modal fade"
        id="deleteModal"
        tabIndex="-1"
        aria-labelledby="deleteModalLabel"
        aria-hidden="true"
      >
        <div className="modal-dialog">
          <div className="modal-content">
            <div className="modal-header">
              <h5 className="modal-title" id="deleteModalLabel">
                Deleting Account '{currUser.username}'
              </h5>
              <button
                type="button"
                className="btn-close"
                data-bs-dismiss="modal"
                aria-label="Close"
              ></button>
            </div>
            <div className="modal-body">
              Are you sure that you want to delete this account? This action
              cannot be reversed.
            </div>
            <div className="modal-footer">
              <button
                type="button"
                className="btn btn-secondary"
                data-bs-dismiss="modal"
              >
                Close
              </button>
              <button
                type="button"
                className="btn btn-danger"
                onClick={removeUser}
                data-bs-dismiss="modal"
              >
                Delete
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

export default Profile;
