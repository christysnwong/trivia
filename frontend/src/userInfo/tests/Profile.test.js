import React from "react";
import { render, screen, fireEvent, waitFor } from "@testing-library/react";
import Profile from "../Profile";
import { MemoryRouter, Router } from "react-router-dom";
import UserProvider from "../../common/UserProvider";

import { createMemoryHistory } from "history";

it("renders without crashing", function () {
  render(
    <UserProvider>
      <Profile />
    </UserProvider>
  );
});

it("matches snapshot", function () {
  const { asFragment } = render(
    <MemoryRouter>
      <UserProvider>
        <Profile />
      </UserProvider>
    </MemoryRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("redirects when clicking edit button", async function () {
  const history = createMemoryHistory();

  render(
    <Router history={history}>
      <UserProvider>
        <Profile />
      </UserProvider>
    </Router>
  );

  const editBtn = screen.getByText(/edit your info/i);
  fireEvent.click(editBtn);

  await waitFor(() => {
    expect(history.location.pathname).toBe("/profile/edit");
  });
});

it("redirects when delete account button", async function () {
  const history = createMemoryHistory();

  render(
    <Router history={history}>
      <UserProvider>
        <Profile />
      </UserProvider>
    </Router>
  );

  const deleteBtn = screen.getAllByText(/delete this account/i)[0];
  fireEvent.click(deleteBtn);

  await waitFor(() => {
    expect(history.location.pathname).toBe("/");
  });
});