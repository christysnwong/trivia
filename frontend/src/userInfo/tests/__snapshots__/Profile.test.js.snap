// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-10 offset-1 col-lg-8 offset-lg-2 col-xxl-6 offset-xxl-3 col-xxxl-4 offset-xxxl-4"
  >
    <h3
      class="mb-5"
    >
      <i
        class="fa-solid fa-address-card me-2"
      />
      Profile
    </h3>
    <div
      class="row mb-3"
    >
      <div
        class="col-4 text-capitalize fw-bold"
      >
        username
      </div>
      <div
        class="col-8"
      >
        testuser
      </div>
    </div>
    <div
      class="row mb-3"
    >
      <div
        class="col-4 text-capitalize fw-bold"
      >
        first_name
      </div>
      <div
        class="col-8"
      >
        Test
      </div>
    </div>
    <div
      class="row mb-3"
    >
      <div
        class="col-4 text-capitalize fw-bold"
      >
        last_name
      </div>
      <div
        class="col-8"
      >
        User
      </div>
    </div>
    <div
      class="row mb-3"
    >
      <div
        class="col-4 text-capitalize fw-bold"
      >
        email
      </div>
      <div
        class="col-8"
      >
        test@test.com
      </div>
    </div>
    <button
      class="btn btn-success mt-4 me-2"
      method="get"
      type="button"
    >
      Edit your info
    </button>
    <button
      class="btn btn-danger mt-4"
      data-bs-target="#deleteModal"
      data-bs-toggle="modal"
      type="button"
    >
      Delete this account
    </button>
    <div
      aria-hidden="true"
      aria-labelledby="deleteModalLabel"
      class="modal fade"
      id="deleteModal"
      tabindex="-1"
    >
      <div
        class="modal-dialog"
      >
        <div
          class="modal-content"
        >
          <div
            class="modal-header"
          >
            <h5
              class="modal-title"
              id="deleteModalLabel"
            >
              Deleting Account 'testuser'
            </h5>
            <button
              aria-label="Close"
              class="btn-close"
              data-bs-dismiss="modal"
              type="button"
            />
          </div>
          <div
            class="modal-body"
          >
            Are you sure that you want to delete this account? This action cannot be reversed.
          </div>
          <div
            class="modal-footer"
          >
            <button
              class="btn btn-secondary"
              data-bs-dismiss="modal"
              type="button"
            >
              Close
            </button>
            <button
              class="btn btn-danger"
              data-bs-dismiss="modal"
              type="button"
            >
              Delete
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
